### PR TITLE
Cache Refactoring

### LINKED ISSUE #2635: HttpCache should not be set via event listeners
| Q | A |
| --- | --- |
| Bug? | no |
| New Feature? | no |
| Sulu Version | Specific version or SHA of a commit |
| Browser Version | Browser name and version |
#### Actual Behavior

Currently the `HttpCache` sets the cache lifetime via event subscribers, and override cache values which might be set in the controller. This is only done for Controller containing a `Structure`.
#### Expected Behavior

The cache values should be set in a more explicit way, and certainly it should be possible to override these values in the controller.
#### Steps to Reproduce
1. Create a custom controller
2. Set a cache life time on the response
3. Use this controller in a template (xml file)
4. See that the cache life time from the response gets overridden
#### Possible Solutions

We should get rid of the cache handlers and especially the subscribers entirely, and revert the cache logic to something more explicit again.

### LINKED ISSUE #3439: Use SnippetReferenceStore in SnippetDataProvider
| Q | A
| --- | ---
| Bug? | yes
| New Feature? | no
| Sulu Version | 1.6
| Browser Version | --

#### Actual Behavior

SnippetReferenceStore isn't used in SnippetDataProvider.

#### Expected Behavior

Snippet references are also stored via SmartContent.

### LINKED ISSUE #3454: Varnish: Add VCL snippets to project
| Q | A
| --- | ---
| Bug? | no
| New Feature? | yes
| Sulu Version | --
| Browser Version | --

#### Actual Behavior

The vcl snippets are listed in the docu.

#### Expected Behavior

Add snippets to project.

### PR DESCRIPTION
**Pull request by trickreich**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #3439, fixes #2635, fixes #3454
| Related issues/PRs | #2635, https://github.com/sulu/sulu-minimal/pull/53
| License | MIT
| Documentation PR | https://github.com/sulu/sulu-docs/pull/348

#### What's in this PR?

Set cache life time headers without listener

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -26,6 +26,12 @@ matrix:
     - env:
         - PHP=true
         - COMPOSER_FLAGS="--prefer-lowest --prefer-dist --no-interaction
+    - php: 5.6
```
Comment (danrot) [inline .travis.yml]: We do not support PHP 5.6 and 7.0 anymore.

File: src/Sulu/Bundle/HttpCacheBundle/Cache/CacheManager.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+namespace Sulu\Bundle\HttpCacheBundle\Cache;
+
+use FOS\HttpCache\ProxyClient\Invalidation\BanCapable;
+use FOS\HttpCacheBundle\CacheManager as FOSCacheManager;
+use Sulu\Component\Webspace\Url\ReplacerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+class CacheManager
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/Cache/CacheManager.php]: Introduce interface?

File: src/Sulu/Bundle/HttpCacheBundle/Cache/HttpCache.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            $this->addSubscriber(new AudienceTargetingListener($this));
+        }
+
+        if (isset($options['debug']) && $options['debug']) {
+            $this->addSubscriber(new DebugListener());
+        }
+    }
+
+    /**
+     * Made public to allow event listeners to do refresh operations.
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/Cache/HttpCache.php]: Is that somehow suggested by the FOSHttpCacheBundle?

File: src/Sulu/Bundle/HttpCacheBundle/CacheLifetime/CacheLifetimeEnhancer.php [hunk 4]
```diff
@@ -52,20 +49,17 @@ class PublicHandler implements HandlerUpdateResponseInterface
      */
     public function __construct(
         CacheLifetimeResolverInterface $cacheLifetimeResolver,
-        $maxAge = 240,
-        $sharedMaxAge = 960,
-        $usePageTtl = true
+        $maxAge,
+        $sharedMaxAge,
+        $usePageTtl
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/CacheLifetime/CacheLifetimeEnhancer.php]: Do we still offer that variable?

File: src/Sulu/Bundle/HttpCacheBundle/Cache/HttpCache.php (hunk 3 above)
Comment (trickreich) [inline src/Sulu/Bundle/HttpCacheBundle/Cache/HttpCache.php]: yes, it is

**Comment by trickreich** [comment]
@danrot I've finished

File: src/Sulu/Bundle/HttpCacheBundle/CacheLifetime/CacheLifetimeEnhancerInterface.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * The cache lifetime enhancer adds cache headers to the response.
+ */
+interface CacheLifetimeEnhancerInterface
+{
+    /**
+     * @param Response $response
+     * @param StructureInterface $structure
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/CacheLifetime/CacheLifetimeEnhancerInterface.php]: remove `@param` annotations and maybe (?) add some explanatory comment.

File: src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -26,38 +40,37 @@ public function getConfigTreeBuilder()
 
         $root
             ->children()
-                ->arrayNode('handlers')
+                ->arrayNode('tags')
+                    ->canBeEnabled()
+                    ->end()
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Configuration.php]: remove one indentation

File: src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -83,6 +94,25 @@ public function getConfigTreeBuilder()
                     ->end()
                 ->end();
 
+        $this->addDebugSection($root);
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Configuration.php]: Any specific reason you have moved that into an own method? If no I would not do it (for the sake of consistency).

File: src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php [hunk 8]
```diff
@@ -28,117 +28,93 @@ class SuluHttpCacheExtension extends Extension implements PrependExtensionInterf
...
+     * Returns boolean if system should cache in current environment.
+     *
+     * @param ContainerInterface $container
      *
-     * @param array $config
-     * @param ContainerBuilder $container
+     * @return bool
      */
-    private function configureStructureCacheHandlers($config, ContainerBuilder $container)
+    protected function shouldCache(ContainerInterface $container): bool
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php]: Make `private`?

File: src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php [hunk 9]
```diff
@@ -28,117 +28,93 @@ class SuluHttpCacheExtension extends Extension implements PrependExtensionInterf
...
+                    $proxyClientAvailable = true;
+                    break;
+                }
+            }
         }
 
-        $container->setParameter('sulu_http_cache.proxy_client.name', $proxyClientName);
-        $container->setAlias('sulu_http_cache.proxy_client', 'sulu_http_cache.proxy_client.' . $proxyClientName);
-    }
+        if ($proxyClientAvailable) {
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php]: Do you think it is necessary to load this configuration only based on the `proxyClient`? I think it doesn't hurt if these services are loaded (correct me if I'm wrong) and there is less place for potential errors if we remove that.

File: src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/InvalidationSubscriber.php [hunk 10]
```diff
@@ -89,17 +82,15 @@ class InvalidationSubscriber implements EventSubscriberInterface
      * @param string $environment - kernel envionment, dev, prod, etc
      */
     public function __construct(
-        HandlerInvalidatePathInterface $pathHandler,
-        HandlerInvalidateStructureInterface $structureHandler,
+        ?CacheManager $cacheManager,
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/InvalidationSubscriber.php]: Instead of making this nullable: Would it make sense to not load this subscriber at all, if caching is not activated? From what I can grasp here, it doesn't do anything, if this is null.

File: src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/TagsSubscriber.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    private $referenceStorePool;
+
+    /**
+     * @var SymfonyResponseTagger
+     */
+    private $symfonyResponseTagger;
+
+    /**
+     * @param ReferenceStorePoolInterface $referenceStorePool
+     * @param SymfonyResponseTagger $symfonyResponseTagger
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/TagsSubscriber.php]: remove php doc

File: src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/TagsSubscriber.php [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    public function addTags()
+    {
+        $this->symfonyResponseTagger->addTags($this->getTags());
+    }
+
+    /**
+     * Merges tags from all registered stores.
+     *
+     * @return array
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/TagsSubscriber.php]: Remove `@return`

File: src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/TagsSubscriber.php [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        return $tags;
+    }
+
+    /**
+     * Returns tags from given store.
+     *
+     * @param string $alias
+     * @param ReferenceStoreInterface $referenceStore
+     *
+     * @return array
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/TagsSubscriber.php]: Remove `@param` and `@return`.

File: src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/CacheLifetime/CacheLifetimeEnhancerTest.php [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+            $this->response->setSharedMaxAge(Argument::any())->shouldNotBeCalled();
+        }
+
+        $this->cacheLifetimeResolver->resolve(Argument::cetera())->willReturn($cacheLifetime);
+        $this->cacheLifetimeEnhancer->enhance($this->response->reveal(), $this->page->reveal());
+    }
+
+    public function testEnhanceSnippet()
+    {
+        $this->cacheLifetimeEnhancer->enhance($this->response->reveal(), $this->snippet->reveal());
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/CacheLifetime/CacheLifetimeEnhancerTest.php]: Shouldn't you test if something has been called here? Doesn't look like this test does assert anything...

File: src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/DependencyInjection/SuluHttpCacheExtensionTest.php [hunk 15]
```diff
@@ -79,91 +81,87 @@ protected function getContainerExtensions()
...
+        $this->assertTrue($this->container->has('sulu_http_cache.cache_lifetime.resolver'));
+        $this->assertTrue($this->container->has('sulu_http_cache.event_subscriber.invalidation'));
+        $this->assertFalse($this->container->has('sulu_http_cache.cache_manager'));
     }
 
-    public function testVarnishConfig()
+    /**
+     * @param bool $debug
+     * @param string $env
+     * @param bool $expected
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/DependencyInjection/SuluHttpCacheExtensionTest.php]: Remove the `@param` annotation

File: src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/EventSubscriber/InvalidationSubscriberTest.php [hunk 16]
```diff
@@ -28,29 +30,22 @@
...
-use Sulu\Component\HttpCache\EventSubscriber\InvalidationSubscriber;
-use Sulu\Component\HttpCache\HandlerInvalidatePathInterface;
-use Sulu\Component\HttpCache\HandlerInvalidateStructureInterface;
 use Sulu\Component\Webspace\Manager\WebspaceManagerInterface;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\RequestStack;
 
 class InvalidationSubscriberTest extends \PHPUnit_Framework_TestCase
 {
+    const DOCUMENT_UUID = '743389e6-2ac5-4673-9835-3e709a27a03d';
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/EventSubscriber/InvalidationSubscriberTest.php]: I think it is not a good idea to introduce constants like this... This test wouldn't fail, if we hard code this uuid in the code. I would use different uuids in every test case.

File: src/Sulu/Bundle/SnippetBundle/Tests/Unit/Content/SnippetDataProviderTest.php [hunk 17]
```diff
@@ -68,7 +75,8 @@ public function setUp()
             $this->snippetQueryBuilder->reveal(),
             $this->nodeHelper->reveal(),
             $this->proxyFactory->reveal(),
-            $this->documentManager->reveal()
+            $this->documentManager->reveal(),
+            $this->referenceStore->reveal()
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Tests/Unit/Content/SnippetDataProviderTest.php]: Maybe also check if the uuids have been correctly added to the reference store in the tests?

File: src/Sulu/Component/CustomUrl/Document/RouteDocument.php [hunk 18]
```diff
@@ -12,15 +12,21 @@
 namespace Sulu\Component\CustomUrl\Document;
 
 use Sulu\Bundle\ContentBundle\Document\RouteDocument as BaseRouteDocument;
+use Sulu\Component\Content\Document\Behavior\WebspaceBehavior;
 
-class RouteDocument extends BaseRouteDocument
+class RouteDocument extends BaseRouteDocument implements WebspaceBehavior
```
Comment (danrot) [inline src/Sulu/Component/CustomUrl/Document/RouteDocument.php]: Why was that necessary?

File: src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * Invalidate custom-url http-cache.
+ */
+class InvalidationSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var CustomUrlManagerInterface
+     */
+    private $manager;
```
Comment (danrot) [inline src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php]: Rename to `customUrlManager`

File: src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * @var null|CacheManager
+     */
+    private $cacheManager;
+
+    /**
+     * @param CustomUrlManagerInterface $manager
+     * @param null|CacheManager $cacheManager
+     */
+    public function __construct(CustomUrlManagerInterface $manager, ?CacheManager $cacheManager)
```
Comment (danrot) [inline src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php]: rename to `$customUrlManager`

File: src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    private $manager;
+
+    /**
+     * @var null|CacheManager
+     */
+    private $cacheManager;
+
+    /**
+     * @param CustomUrlManagerInterface $manager
+     * @param null|CacheManager $cacheManager
```
Comment (danrot) [inline src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php]: Remove php doc

File: src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+            Events::REMOVE => ['invalidateDocumentBeforeRemoving', 1024],
+        ];
+    }
+
+    /**
+     * Invalidate custom-urls for persisted pages.
+     *
+     * @param PersistEvent $event
+     */
+    public function handlePersist(PersistEvent $event)
```
Comment (danrot) [inline src/Sulu/Component/CustomUrl/Document/Subscriber/InvalidationSubscriber.php]: Do we need that method? I think now that you correctly invalidate the cache on publish it's not needed anymore (I think it is even executed twice now...)

File: src/Sulu/Component/CustomUrl/Document/RouteDocument.php (hunk 18 above)
Comment (trickreich) [inline src/Sulu/Component/CustomUrl/Document/RouteDocument.php]: Because i need the webspace of the document

File: src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php (hunk 9 above)
Comment (trickreich) [inline src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php]: We let this the way

File: src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/InvalidationSubscriber.php (hunk 10 above)
Comment (trickreich) [inline src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/InvalidationSubscriber.php]: As discussed: We do not change this.

File: src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/CacheLifetime/CacheLifetimeEnhancerTest.php (hunk 14 above)
Comment (trickreich) [inline src/Sulu/Bundle/HttpCacheBundle/Tests/Unit/CacheLifetime/CacheLifetimeEnhancerTest.php]: Yes, you are right

File: src/Sulu/Bundle/SnippetBundle/Tests/Unit/Content/SnippetDataProviderTest.php (hunk 17 above)
Comment (trickreich) [inline src/Sulu/Bundle/SnippetBundle/Tests/Unit/Content/SnippetDataProviderTest.php]: Yes, is implemented

**Comment by trickreich** [comment]
#3439 Is also implemented here

**Comment by danrot** [comment]
You should add `fixes #3439` to the description then, so that the issue gets closed when we merge this. Already did that for you now.

File: src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+        // posthandle
+        $audienceTargetingCacheListener->postHandle($this->getCacheEvent($httpCache, $request, $response));
+
+        // check if headers are set
+        $this->assertEquals(0, $response->getMaxAge());
+
+        // check if both cookies are set
+        $this->assertCount(2, $response->headers->getCookies());
+
+        // check for cookie one
```
Comment (danrot) [inline src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php]: `check for target group cookie` or remove

File: src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+        $this->assertEquals(
+            'TARGET_GROUP_1',
+            $response->headers->getCookies()[0]->getValue()
+        );
+        $this->assertEquals(
+            AudienceTargetingCacheListener::TARGET_GROUP_COOKIE_LIFETIME,
+            $response->headers->getCookies()[0]->getExpiresTime()
+        );
+
+        // check for cookie two
```
Comment (danrot) [inline src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php]: `check for session cookie` or remove

File: src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+
+    public function testHandleWithCorrectCookies()
+    {
+        $request = $this->getRequest(true);
+        $response = $this->getResponse();
+        $httpCache = $this->getHttpCache();
+
+        $audienceTargetingCacheListener = new AudienceTargetingCacheListener();
+
+        // prehandle
```
Comment (danrot) [inline src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php]: Please remove

File: src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+        $audienceTargetingCacheListener->preHandle($this->getCacheEvent($httpCache, $request, $response));
+
+        $this->assertAttributeEquals(true, 'hadValidTargetGroupCookie', $audienceTargetingCacheListener);
+        $this->assertEmpty($response->headers->getCookies());
+
+        // posthandle
```
Comment (danrot) [inline src/Sulu/Bundle/AudienceTargetingBundle/Tests/Unit/EventListener/AudienceTargetingCacheListenerTest.php]: Please remove

File: src/Sulu/Bundle/HttpCacheBundle/Cache/CacheManager.php [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+
+            $path = $this->replacer->replaceHost($path, $this->requestHost);
+        }
+
+        $this->cacheManager->invalidatePath($path, $headers);
+
+        if (!$this->cacheManager->supports(FOSCacheManager::REFRESH)) {
+            return;
+        }
+        $this->cacheManager->refreshPath($path, $headers);
```
Comment (danrot) [inline src/Sulu/Bundle/HttpCacheBundle/Cache/CacheManager.php]: Move that line into `if` (if there are more options like this in the future it wouldn't be able to return early)

File: src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php [hunk 28]
```diff
@@ -69,7 +70,15 @@ protected function renderStructure(
                 );
             }
 
-            return new Response($content);
+            // create response
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php]: Please remove

File: src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php [hunk 29]
```diff
@@ -69,7 +70,15 @@ protected function renderStructure(
...
+            $response = new Response($content);
+
+            if (!$preview && $this->getCacheTimeLifeEnhancer()) {
+                // use of cache time life enhancer
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php]: Please remove

File: src/Sulu/Bundle/AudienceTargetingBundle/EventListener/AudienceTargetingCacheListener.php [hunk 30]
```diff
@@ -37,80 +36,43 @@ class HttpCache extends AbstractHttpCache
...
-
-        if ($this->hasAudienceTargeting && !$hadValidTargetGroupCookie) {
+        if (!$this->hadValidTargetGroupCookie) {
             $this->setTargetGroupCookie($response, $request);
         }
 
-        if ($this->hasAudienceTargeting && in_array(static::TARGET_GROUP_HEADER, $response->getVary())) {
+        if (in_array(static::TARGET_GROUP_HEADER, $response->getVary())) {
             $response->setMaxAge(0);
             $response->setSharedMaxAge(0);
```
Comment (danrot) [inline src/Sulu/Bundle/AudienceTargetingBundle/EventListener/AudienceTargetingCacheListener.php]: This seems to be overridden again later. See the following screenshot:
image

Although there is a vary header, the shared max age is still set to 240.

**Comment by danrot** [comment]
@trickreich Found and fixed the error with the 500 I had 😄

File: src/Sulu/Bundle/AudienceTargetingBundle/Resources/varnish/sulu.vcl [hunk 31]
```diff
@@ -13,12 +13,12 @@ sub sulu_audience_targeting_recv {
...
     } elseif (req.restarts == 0) {
-        set req.http.X-Sulu-Original-Url = req.url;
+        set req.http.X-Forwarded-Url = req.url;
         set req.http.X-Sulu-Target-Group = regsub(req.http.Cookie, ".*_svtg=([^;]+).*", "\1");
         set req.url = "/_sulu_target_group";
     } elseif (req.restarts > 0) {
-        set req.url = req.http.X-Sulu-Original-Url;
-        unset req.http.X-Sulu-Original-Url;
+        set req.url = req.http.X-Forwarded-Url;
+        unset req.http.X-Forwarded-Url ;
```
Comment (chirimoya) [inline src/Sulu/Bundle/AudienceTargetingBundle/Resources/varnish/sulu.vcl]: remove space

**Comment by alexander-schranz** [comment]
think this PR will also make #2829 unnecessary? /cc @trickreich

**Comment by danrot** [comment]
We should remember that the stuff @wachterjohannes has changed (#3779 and #3796) should be still working after the rebase.

**Comment by danrot** [comment]
Now there is also PR #3778. Once you've rebased I'll have another look and test it, and Johannes or Alex will test it, so that this can get merged quite soon.

**Comment by wachterjohannes** [comment]
@danrot the strange behaviour was introduced by the `no-cache` handling - @trickreich has disable this in the last commit - so this should be good now. maybe test it and for me it could be merged then
